src/pointer.rs

```diff
@@ -0,0 +1,164 @@
+//! "Pointer"s into a syntax tree.
+//!
+//! These pointers are "trace"s down a syntax tree, e.g. `[4, 2, 0]` says to
+//! take the 4th child, the 2nd child, and then the 0th child. To reduce the
+//! size and allocation requirements of syntax pointers, they perform a "small
+//! vector" optimization, storing the trace inline if it takes 15 or fewer
+//! bytes to represent. Additionally, a custom variable-length encoding is
+//! used to store pointers compactly while still supporting large trees.
+//!
+//! The actual trace encoding is as follows:
+//!
+//! - For child indices < 240 (0xF0),
+//!       store the index as a byte, as-is.
+//! - For child indices >= 240 (0xF0),
+//!       store the index as two bytes, 0xF_ 0x__.
+//! - Indices > 4095 (0x0FFF) are unrepresentable.
+
+use smallvec::SmallVec;
+use std::{convert::TryInto, fmt, mem, slice};
+
+#[cfg(not(any(target_pointer_width = "32", target_pointer_width = "64")))]
+compile_error!("Rowan currently only works on 32 or 64 bit targets");
+
+/// A pointer trace in some syntax tree to a specific node.
+#[repr(C)]
+#[cfg_attr(target_pointer_width = "32", repr(align(4)))]
+#[cfg_attr(target_pointer_width = "64", repr(align(8)))]
+pub struct SyntaxPtr {
+    data: SyntaxPtrData,
+    len: u8,
+}
+
+#[repr(C, packed)]
+union SyntaxPtrData {
+    inline: [u8; 15],
```

Comment:
Why is this only 15 bytes, when `outline` is 16 bytes on 64bit systems?

Reply:
It's not: I only store `*mut u8`, which is one pointer (8 bytes). The data is 15 bytes because the 16th byte is used as a length.